esp-hal/src/lib.rs

```diff
@@ -607,6 +607,17 @@ impl<T, const SIZE: usize> FlashSafeDma<T, SIZE> {
     }
 }
 
+/// Trait implemented by drivers which allow the user to set an
+/// [InterruptHandler]
+pub trait InterruptConfigurable: private::Sealed {
```

Comment:
As is, I'm not sure how I would consume this trait as a user, but with two certain additions, one could make useful abstractions with it.

One is to add an associated type that represent the interrupts bit. Either a boolean struct like in Pcnt driver or an EnumSet like in the Spi driver.

The second is to expose methods to access the INT_CLR, INT_ENA, INT_ST, INT_RAW registers.

Idk if it makes sense for this PR or as a future enhancement but I think one can write a generic `Future` on top of this. (Barring the awkward case when multiple interrupt handlers share the same INT_* registers like `SYS_TIMER`)

Reply:
I also had some doubts initially about the trait ( https://github.com/esp-rs/esp-hal/issues/1489#issuecomment-2068566303 ) but I agree that with some additions it could become more useful

Looking at the amount of changed files I probably would say that's something for another PR since getting rid of the handler from the ctors is already a good thing on its own (since in most cases there is an annoying `None` passed)

Reply:
Sorry I didn't fully explain my position in the original issue, but I don't see the trait in its current form as final, I just wanted a place to start to have a consistent API and something that we can start to abstract over (I didn't like that we had two ways to set interrupts).

I think your suggestions would be great in a future PR :)